components/analytics/bot-performance-chart.tsx

```
@@ -0,0 +1,373 @@
+    })
+  }
+
+  // Data transformation for stacked percentage chart
+  const getPercentageData = (): PercentageDataItem[] => {
+    return chartData.map(day => {
+      const platformData: Record<string, number> = {}
+      const total = day.totalBots || 1 // Avoid division by zero
+
+      allPlatforms.forEach(platform => {
+        platformData[`${platform}Percentage`] = ((day[platform] as number) || 0) / total * 100
+      })
```

Biome (1.9.4)

[error] Prefer for...of instead of forEach.

forEach may lead to performance issues when working with large arrays. When combined with functions like filter or map, this causes multiple iterations over the same type.

(lint/complexity/noForEach)